Sources/AWSSDKSwiftCore/HTTP/AsyncHTTPClient.swift

```diff
@@ -18,19 +18,71 @@ import NIO
+    private func writeToStreamWriter(
+        writer: HTTPClient.Body.StreamWriter,
+        size: Int?,
+        on eventLoop: EventLoop,
+        stream: @escaping (EventLoop)->EventLoopFuture<ByteBuffer>) -> EventLoopFuture<Void> {
+        let promise = eventLoop.makePromise(of: Void.self)
+
+        func _writeToStreamWriter(_ amountLeft: Int?) {
+            // get byte buffer from closure, write to StreamWriter, if there are still bytes to write then call
+            // _writeToStreamWriter again.
+            _ = stream(eventLoop)
+                .map { (byteBuffer)->() in
+                    // if no amount was set and the byte buffer has no readable bytes then this is assumed to mean
+                    // there will be no more data
+                    if amountLeft == nil && byteBuffer.readableBytes == 0 {
+                        promise.succeed(())
+                        return
+                    }
+                    let newAmountLeft = amountLeft.map { $0 - byteBuffer.readableBytes }
+                    _ = writer.write(.byteBuffer(byteBuffer)).flatMap { ()->EventLoopFuture<Void> in
+                        if let newAmountLeft = newAmountLeft {
+                            if newAmountLeft == 0 {
+                                promise.succeed(())
+                            } else if newAmountLeft < 0 {
+                                promise.fail(AWSClient.ClientError.tooMuchData)
+                            } else {
+                                _writeToStreamWriter(newAmountLeft)
+                            }
+                        } else {
+                            _writeToStreamWriter(nil)
+                        }
+                        return promise.futureResult
+                    }.cascadeFailure(to: promise)
+            }.cascadeFailure(to: promise)
+        }
+        _writeToStreamWriter(size)
+        return promise.futureResult
+    }
+
     public func execute(request: AWSHTTPRequest, timeout: TimeAmount, on eventLoop: EventLoop?) -> EventLoopFuture<AWSHTTPResponse> {
         if let eventLoop = eventLoop {
             precondition(self.eventLoopGroup.makeIterator().contains { $0 === eventLoop }, "EventLoop provided to AWSClient must be part of the HTTPClient's EventLoopGroup.")
-        }
+        }        
+        let eventLoop = eventLoop ?? eventLoopGroup.next()
         let requestBody: AsyncHTTPClient.HTTPClient.Body?
-        if let body = request.body {
-            requestBody = .byteBuffer(body)
-        } else {
+        var requestHeaders = request.headers
+        
+        switch request.body?.payload {
+        case .some(.byteBuffer(let byteBuffer)):
+            requestBody = .byteBuffer(byteBuffer)
+        case .some(.stream(let size, let closure)):
+            // add "Transfer-Encoding" header if streaming with unknown size
+            if size == nil {
+                requestHeaders.add(name: "Transfer-Encoding", value: "chunked")
```

Comment:
awesome. if we do this though we need to add some kind of magic headers between the chunks. i think those are missing right now.

Reply:
With an unsigned-payload stream you don't need the magic headers. There are a number of services that only accept unsigned-payload streams, also S3 does accept unsigned-payloads. I have already added an issue to do a signed stream #248 at a later date.

Also if you don't provide a size it isn't possible to do a signed streamed upload as the final content-size is required.

Reply:
Does ahc add the sizes and \r\n between the chunks as defined here:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding

Reply:
@fabianfett good point, will need to check. So far I've only checked this with non-chunked uploads.

Reply:
yep it failed doing an S3.PutObject. This is something I would have thought was really up to AHC, but then they expect you to set the "chunked" header. I'll ask them

Reply:
https://github.com/swift-server/async-http-client/issues/212

Reply:
By the way AWS have their own version of this called "aws-chunked" which includes signing information in it. https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-streaming.html